res/controllers/libpcon.js

```diff
@@ -0,0 +1,378 @@
+// eslint-disable-next-line no-var, no-redeclare
+var pcon = {};
+
+// /**
+//  * this converts a typed array from one to the other, respecting the views offset and length.
+//  * in theory, this is similar to a C++ reinterpret_cast
+//  * @param {Function} cls constructor of a TypedArray to cast to
+//  * @param {TypedArray} from TypedArray
+//  * @returns {cls} the type being cast too
+//  */
+// pcon.castViewTo = function(cls, from) {
+//     if (from instanceof cls) {
+//         return from;
+//     }
+//     if (from instanceof ArrayBuffer) {
+//         return new cls(from);
+//     }
+//     if (from instanceof Array) {
+//         return pcon.castViewTo(cls, new Uint8Array(from));
+//     }
+//     if (cls === DataView) {
+//         return new DataView()
+//     }
+//     const div =  ? 1 : cls.BYTES_PER_ELEMENT;
+//     return new cls(from.buffer, from.byteOffset, Math.trunc((from.byteLength) / div));
+// };
+
+pcon.util = {
+    isByte: function(size) {
+        return size >= 0x00 && size <= 0xFF;
+    },
+    isASCII: function(byte) {
+        return byte >= 0x20 && byte <= 0x7E;
+    },
+    chunk: function(arr, size) {
+        const ret = [];
+        for (let chunk = 0; chunk <  arr.length;) {
+            ret.push(Array.from(arr.slice(chunk, chunk+=size)));
+        }
+        return ret;
+    },
+    fromHexStream: function(str) {
+        return new Uint8Array(pcon.util.chunk(str, 2).map(byte => Number.parseInt(byte.join(""), 16)));
+    }
+};
+
+pcon.debug = {
+    hexDump: function(buff) {
+        return pcon.util.chunk(buff, 16).map(chunk => {
+            const asciiRep = chunk.map(byte => pcon.util.isASCII(byte) ? String.fromCharCode(byte) : "•").join("");
+            const bufHex = chunk.map(byte => pcon.util.isByte(byte) ? byte.toString(16).padStart(2, "0") : "  ").join(" ");
+            const charsPerByte = 3;
+            const pad = Math.min(buff.length, 16);
+            const strBuffer = `${bufHex.slice(0, 8*charsPerByte)} ${bufHex.slice(8*charsPerByte, 16*charsPerByte)}`.padEnd(pad*charsPerByte);
+            return `${strBuffer} | ${asciiRep}`;
+        })
+            .join("\n");
+    }
+};
+
+
+pcon.makeTLV = function(type, data) {
+    console.assert(pcon.util.isByte(type));
+    console.assert(pcon.util.isByte(data.length + 2));
+    return [type, data.length + 2].concat(data);
+};
+/**
+ *
+ * @param {Uint8Array} data ArrayBuffer containing the start of the TLV, along with trailing data
+ * @returns {Uint8Array} TODO define
+ */
+pcon.readTLV = function(data) {
+    if (data.length < 2) {
+        return undefined;
+    }
+    const type = data[0];
+    const length = data[1];
+
+    console.assert(length <= data.length);
+    const value = data.slice(2, length);
+
+    console.debug(type.toString(16));
+    console.debug(length.toString(16));
+    console.debug(value);
```

Comment:
It would be great to either add some message prefix on `console.debug` so we don't get logs with a single word spam which is hard to link back to where it comes from. Alternative, we could the `controllerDebug` bool to hide this if not enabled.